### ISSUE TITLE
[v2.6] `permission denied` error when relation joins before leader-elected

### ISSUE THREAD

**Issue by sed-i** [canonical/tls-certificates-interface#58]
I deployed alertmanager and quickly related it to self-signed-certificates.

From the event sequence we can see that relation events for alertmanager fired before leader-elected:
```
unit-am-0: 21:00:07.168 DEBUG unit.am/0.juju-log Emitting Juju event install.
unit-ca-0: 21:00:07.913 DEBUG unit.ca/0.juju-log certificates:10: Emitting Juju event certificates_relation_joined.
unit-am-0: 21:00:08.042 DEBUG unit.am/0.juju-log certificates:10: Emitting Juju event certificates_relation_created.
unit-ca-0: 21:00:08.303 DEBUG unit.ca/0.juju-log certificates:10: Emitting Juju event certificates_relation_changed.
unit-am-0: 21:00:08.592 DEBUG unit.am/0.juju-log replicas:9: Emitting Juju event replicas_relation_created.
unit-ca-0: 21:00:09.250 DEBUG unit.ca/0.juju-log certificates:10: Emitting Juju event certificates_relation_changed.
unit-ca-0: 21:00:09.259 DEBUG unit.ca/0.juju-log certificates:10: Emitting custom event <CertificateCreationRequestEvent via SelfSignedCertificatesCharm/TLSCertificatesProvidesV2[certificates]/on/certificate_creation_request[107]>.
unit-am-0: 21:00:09.395 DEBUG unit.am/0.juju-log Emitting Juju event leader_elected.
```

As a result, self-signed-certificates charm is in error state (originating from the tls-certificates library):
```
unit-ca-0: 21:02:01.919 DEBUG unit.ca/0.juju-log certificates:10: ops 2.4.1 up and running.
unit-ca-0: 21:02:01.939 DEBUG unit.ca/0.juju-log certificates:10: Emitting Juju event certificates_relation_changed.
unit-ca-0: 21:02:01.948 DEBUG unit.ca/0.juju-log certificates:10: Emitting custom event <CertificateCreationRequestEvent via SelfSignedCertificatesCharm/TLSCertificatesProvidesV2[certificates]/on/certificate_creation_request[111]>.
unit-ca-0: 21:02:06.229 INFO unit.ca/0.juju-log certificates:10: Generated certificate for relation 10
unit-ca-0: 21:02:06.247 ERROR unit.ca/0.juju-log certificates:10: Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-ca-0/charm/venv/ops/model.py", line 2693, in _run
    result = subprocess.run(args, **kwargs)  # type: ignore
  File "/usr/lib/python3.10/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '('/var/lib/juju/tools/unit-ca-0/relation-get', '-r', '6', '-', 'ca', '--app', '--format=json')' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-ca-0/charm/./src/charm.py", line 200, in <module>
    main(SelfSignedCertificatesCharm)
  File "/var/lib/juju/agents/unit-ca-0/charm/venv/ops/main.py", line 441, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-ca-0/charm/venv/ops/main.py", line 149, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-ca-0/charm/venv/ops/framework.py", line 344, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-ca-0/charm/venv/ops/framework.py", line 833, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-ca-0/charm/venv/ops/framework.py", line 922, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-ca-0/charm/lib/charms/tls_certificates_interface/v2/tls_certificates.py", line 1095, in _on_relation_changed
    self._revoke_certificates_for_which_no_csr_exists(relation_id=event.relation.id)
  File "/var/lib/juju/agents/unit-ca-0/charm/lib/charms/tls_certificates_interface/v2/tls_certificates.py", line 1129, in _revoke_certificates_for_which_no_csr_exists
    self.remove_certificate(certificate=certificate["certificate"])
  File "/var/lib/juju/agents/unit-ca-0/charm/lib/charms/tls_certificates_interface/v2/tls_certificates.py", line 1054, in remove_certificate
    self._remove_certificate(certificate=certificate, relation_id=certificate_relation.id)
  File "/var/lib/juju/agents/unit-ca-0/charm/lib/charms/tls_certificates_interface/v2/tls_certificates.py", line 961, in _remove_certificate
    provider_relation_data = _load_relation_data(relation.data[self.charm.app])
  File "/var/lib/juju/agents/unit-ca-0/charm/lib/charms/tls_certificates_interface/v2/tls_certificates.py", line 597, in _load_relation_data
    for key in raw_relation_data:
  File "/var/lib/juju/agents/unit-ca-0/charm/venv/ops/model.py", line 675, in __iter__
    return iter(self._data)
  File "/var/lib/juju/agents/unit-ca-0/charm/venv/ops/model.py", line 662, in _data
    data = self._lazy_data = self._load()
  File "/var/lib/juju/agents/unit-ca-0/charm/venv/ops/model.py", line 1393, in _load
    return self._backend.relation_get(self.relation.id, self._entity.name, self._is_app)
  File "/var/lib/juju/agents/unit-ca-0/charm/venv/ops/model.py", line 2774, in relation_get
    raw_data_content = self._run(*args, return_output=True, use_json=True)
  File "/var/lib/juju/agents/unit-ca-0/charm/venv/ops/model.py", line 2695, in _run
    raise ModelError(e.stderr)
ops.model.ModelError: ERROR permission denied (unauthorized access)
```

**Comment by sed-i** [comment]
Not sure if it's a juju bug. @benhoyt ?

**Comment by sed-i** [comment]
Just noticed that the cert provider (self-signed-certificates) does `relation-get', '-r', '6'`, but from the log we see that the certificates relation id is 10.
Not sure what's going on.

**Comment by benhoyt** [comment]
Yeah, seems like something (probably the charm lib, but perhaps ops?) is using the wrong relation ID, leading to the permission error.

**Comment by gruyaume** [comment]
Fixed by PR #70
